{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "pycharm": {
     "is_executing": true,
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd  \n",
    "import numpy as np  \n",
    "import os  \n",
    "import sys  \n",
    "import collections\n",
    "from itertools import chain\n",
    "import json\n",
    "from IPython.display import display, HTML\n",
    "import csv\n",
    "import glob\n",
    "import h5py\n",
    "import re\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "pycharm": {
     "is_executing": true,
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "pd.set_option('display.max_rows', 5000)\n",
    "pd.set_option('display.max_colwidth', 0)\n",
    "pd.options.display.float_format = lambda x: '{:,.2f}'.format(x) if x > 1 else '{:,.2%}'.format(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "from IPython.core.interactiveshell import InteractiveShell\n",
    "InteractiveShell.ast_node_interactivity = \"all\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "top_metrics_file = None\n",
    "dict_features ={}\n",
    "top_metrics_file = 'top_metrics_for_tbl.csv'\n",
    "dict_features = {row[0] : row[1] for _, row in pd.read_csv(top_metrics_file).iterrows()}\n",
    "df_features = pd.read_csv(top_metrics_file)\n",
    "list_metrics = list(set(df_features['metric'] ))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "# Get input H5 file\n",
    "input_h5_file = []\n",
    "input_h5_file = 'input_for_html_report.h5'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get base_file_name\n",
    "with open('base_file_name.txt', 'r') as f:\n",
    "    input_base_file_name = f.readline().rstrip()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "# get Keys within the H5 file\n",
    "f = h5py.File(input_h5_file, 'r')\n",
    "list_h5_df = []\n",
    "list_keys = list(f.keys())\n",
    "hist_list_keys =  [i for i in list_keys if i.startswith('histogram_')]\n",
    "tbl_list_keys = pd.DataFrame(list(set(list_keys) - set(hist_list_keys)))\n",
    "tbl_list_keys.columns = ['metric']\n",
    "del(list_keys)\n",
    "\n",
    "# create table merging top required metrics to display and input provided\n",
    "# ....................\n",
    "tbl_top_values = pd.merge(df_features,tbl_list_keys,on='metric',how='inner')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>key</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Sample name</td>\n",
       "      <td>baseline_L3511_016796_HG001_rep1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           key                             value\n",
       "0  Sample name  baseline_L3511_016796_HG001_rep1"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "to_print = pd.DataFrame(data = {'key' : 'Sample name', 'value' : input_base_file_name}, index = [0])\n",
    "to_print"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<b></b>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<hr/>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<h2 style=\"font-size:20px;\">Summary View: Main Metrics</h2>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<hr/>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<b></b>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "HTML(\"<b>\"+\"\"+\"</b>\")\n",
    "HTML(\"<hr/>\")\n",
    "HTML(\"<h2 style=\\\"font-size:20px;\\\">\"+\"Summary View: Main Metrics\"+\"</h2>\")\n",
    "HTML(\"<hr/>\")\n",
    "HTML(\"<b>\"+\"\"+\"</b>\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>key</th>\n",
       "      <th>value</th>\n",
       "      <th>metric</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TOTAL_READS</td>\n",
       "      <td>614332000</td>\n",
       "      <td>AlignmentSummaryMetrics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PCT_PF_READS</td>\n",
       "      <td>99.77%</td>\n",
       "      <td>AlignmentSummaryMetrics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PCT_PF_READS_ALIGNED</td>\n",
       "      <td>99.13%</td>\n",
       "      <td>AlignmentSummaryMetrics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PF_MISMATCH_RATE</td>\n",
       "      <td>20.15%</td>\n",
       "      <td>AlignmentSummaryMetrics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PF_INDEL_RATE</td>\n",
       "      <td>0.76%</td>\n",
       "      <td>AlignmentSummaryMetrics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MEAN_READ_LENGTH</td>\n",
       "      <td>224.68</td>\n",
       "      <td>AlignmentSummaryMetrics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MEAN_ALIGNED_READ_LENGTH</td>\n",
       "      <td>220.49</td>\n",
       "      <td>AlignmentSummaryMetrics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PF_BASES</td>\n",
       "      <td>137717547905</td>\n",
       "      <td>QualityYieldMetricsFlow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>PF_Q30_BASES</td>\n",
       "      <td>91026891229</td>\n",
       "      <td>QualityYieldMetricsFlow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>GC_NC_0_19</td>\n",
       "      <td>52.85%</td>\n",
       "      <td>GcBiasSummaryMetrics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>GC_NC_80_100</td>\n",
       "      <td>1.82</td>\n",
       "      <td>GcBiasSummaryMetrics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>MEAN_COVERAGE</td>\n",
       "      <td>41.79</td>\n",
       "      <td>RawWgsMetrics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>PCT_20X</td>\n",
       "      <td>94.83%</td>\n",
       "      <td>RawWgsMetrics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>FOLD_90_BASE_PENALTY</td>\n",
       "      <td>1.67</td>\n",
       "      <td>RawWgsMetrics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Genome___percentile_5</td>\n",
       "      <td>21.51</td>\n",
       "      <td>stats_coverage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Exome (WG)___percentile_5</td>\n",
       "      <td>25.92</td>\n",
       "      <td>stats_coverage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Simple repeat___percentile_5</td>\n",
       "      <td>45.17%</td>\n",
       "      <td>stats_coverage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>PERCENT_DUPLICATION</td>\n",
       "      <td>6.57%</td>\n",
       "      <td>DuplicationMetrics</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             key         value                   metric\n",
       "0   TOTAL_READS                   614332000     AlignmentSummaryMetrics\n",
       "1   PCT_PF_READS                 99.77%         AlignmentSummaryMetrics\n",
       "2   PCT_PF_READS_ALIGNED         99.13%         AlignmentSummaryMetrics\n",
       "3   PF_MISMATCH_RATE             20.15%         AlignmentSummaryMetrics\n",
       "4   PF_INDEL_RATE                0.76%          AlignmentSummaryMetrics\n",
       "5   MEAN_READ_LENGTH             224.68         AlignmentSummaryMetrics\n",
       "6   MEAN_ALIGNED_READ_LENGTH     220.49         AlignmentSummaryMetrics\n",
       "7   PF_BASES                      137717547905  QualityYieldMetricsFlow\n",
       "8   PF_Q30_BASES                  91026891229   QualityYieldMetricsFlow\n",
       "9   GC_NC_0_19                   52.85%         GcBiasSummaryMetrics   \n",
       "10  GC_NC_80_100                 1.82           GcBiasSummaryMetrics   \n",
       "11  MEAN_COVERAGE                41.79          RawWgsMetrics          \n",
       "12  PCT_20X                      94.83%         RawWgsMetrics          \n",
       "13  FOLD_90_BASE_PENALTY         1.67           RawWgsMetrics          \n",
       "14  Genome___percentile_5        21.51          stats_coverage         \n",
       "15  Exome (WG)___percentile_5    25.92          stats_coverage         \n",
       "16  Simple repeat___percentile_5 45.17%         stats_coverage         \n",
       "17  PERCENT_DUPLICATION          6.57%          DuplicationMetrics     "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "separator = '___'\n",
    "i_value= np.nan \n",
    "to_print = pd.DataFrame()\n",
    "\n",
    "tbl_top = pd.DataFrame()\n",
    "for temp_metric in tbl_top_values['metric'].unique():\n",
    "    list_top_tbl = sub_top_tbl = pd.DataFrame()\n",
    "    sub_top_tbl =  tbl_top_values[tbl_top_values['metric'] == temp_metric]\n",
    "    df_h5_tbl = pd.read_hdf(input_h5_file, temp_metric).T\n",
    "    df_h5_tbl = df_h5_tbl.reset_index()\n",
    "    \n",
    "    # stats_coverage is a multiindex dataframe    \n",
    "    if (temp_metric.startswith('stats') ):\n",
    "        df_h5_tbl['metric'] = df_h5_tbl['level_0'] + separator + df_h5_tbl['level_1']\n",
    "        df_h5_tbl = df_h5_tbl.drop(columns=['level_0','level_1']).copy()\n",
    "        df_h5_tbl.columns = ['value','key']\n",
    "        df_h5_tbl = df_h5_tbl[['key', 'value']]        \n",
    "        list_top_tbl = pd.merge(df_h5_tbl, sub_top_tbl ,on='key',how='inner')\n",
    "        to_print = to_print.append(list_top_tbl)\n",
    "        \n",
    "    else:\n",
    "        df_h5_tbl.columns = ['key','value']        \n",
    "        list_top_tbl = pd.merge(df_h5_tbl, sub_top_tbl ,on='key',how='inner')\n",
    "        to_print = to_print.append(list_top_tbl)\n",
    "if type(to_print.loc[to_print.key == 'PCT_PF_READS', 'value'].values[0]) is int:\n",
    "    to_print.loc[to_print.key == 'PCT_PF_READS', 'value'] = '100%'\n",
    "to_print.reset_index().drop(columns = 'index')\n",
    "\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<b></b>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<hr/>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<h2 style=\"font-size:20px;\">Detailed View: All Metrics</h2>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<hr/>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<b></b>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HTML(\"<b>\"+\"\"+\"</b>\")\n",
    "HTML(\"<hr/>\")\n",
    "HTML(\"<h2 style=\\\"font-size:20px;\\\">\"+\"Detailed View: All Metrics\"+\"</h2>\")\n",
    "HTML(\"<hr/>\")\n",
    "HTML(\"<b>\"+\"\"+\"</b>\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<br><br><b>Metric type: AlignmentSummaryMetrics</b><br>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>key</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CATEGORY</td>\n",
       "      <td>UNPAIRED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TOTAL_READS</td>\n",
       "      <td>614332000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PF_READS</td>\n",
       "      <td>612938364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PCT_PF_READS</td>\n",
       "      <td>99.77%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PF_NOISE_READS</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PF_READS_ALIGNED</td>\n",
       "      <td>607635638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PCT_PF_READS_ALIGNED</td>\n",
       "      <td>99.13%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PF_ALIGNED_BASES</td>\n",
       "      <td>134730406449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>PF_HQ_ALIGNED_READS</td>\n",
       "      <td>541239186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>PF_HQ_ALIGNED_BASES</td>\n",
       "      <td>120730619622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>PF_HQ_ALIGNED_Q20_BASES</td>\n",
       "      <td>106072750701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>PF_HQ_MEDIAN_MISMATCHES</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>PF_MISMATCH_RATE</td>\n",
       "      <td>20.15%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>PF_HQ_ERROR_RATE</td>\n",
       "      <td>20.24%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>PF_INDEL_RATE</td>\n",
       "      <td>0.76%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>MEAN_READ_LENGTH</td>\n",
       "      <td>224.68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>SD_READ_LENGTH</td>\n",
       "      <td>59.88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>MEDIAN_READ_LENGTH</td>\n",
       "      <td>222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>MAD_READ_LENGTH</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>MIN_READ_LENGTH</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>MAX_READ_LENGTH</td>\n",
       "      <td>444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>MEAN_ALIGNED_READ_LENGTH</td>\n",
       "      <td>220.49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>READS_ALIGNED_IN_PAIRS</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>PCT_READS_ALIGNED_IN_PAIRS</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>PF_READS_IMPROPER_PAIRS</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>PCT_PF_READS_IMPROPER_PAIRS</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>BAD_CYCLES</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>STRAND_BALANCE</td>\n",
       "      <td>50.02%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>PCT_CHIMERAS</td>\n",
       "      <td>1.58%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>PCT_ADAPTER</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>PCT_SOFTCLIP</td>\n",
       "      <td>1.15%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>PCT_HARDCLIP</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>AVG_POS_3PRIME_SOFTCLIP_LENGTH</td>\n",
       "      <td>60.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>SAMPLE</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>LIBRARY</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>READ_GROUP</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               key         value\n",
       "0   CATEGORY                        UNPAIRED    \n",
       "1   TOTAL_READS                     614332000   \n",
       "2   PF_READS                        612938364   \n",
       "3   PCT_PF_READS                   99.77%       \n",
       "4   PF_NOISE_READS                  0           \n",
       "5   PF_READS_ALIGNED                607635638   \n",
       "6   PCT_PF_READS_ALIGNED           99.13%       \n",
       "7   PF_ALIGNED_BASES                134730406449\n",
       "8   PF_HQ_ALIGNED_READS             541239186   \n",
       "9   PF_HQ_ALIGNED_BASES             120730619622\n",
       "10  PF_HQ_ALIGNED_Q20_BASES         106072750701\n",
       "11  PF_HQ_MEDIAN_MISMATCHES         40          \n",
       "12  PF_MISMATCH_RATE               20.15%       \n",
       "13  PF_HQ_ERROR_RATE               20.24%       \n",
       "14  PF_INDEL_RATE                  0.76%        \n",
       "15  MEAN_READ_LENGTH               224.68       \n",
       "16  SD_READ_LENGTH                 59.88        \n",
       "17  MEDIAN_READ_LENGTH              222         \n",
       "18  MAD_READ_LENGTH                 45          \n",
       "19  MIN_READ_LENGTH                 50          \n",
       "20  MAX_READ_LENGTH                 444         \n",
       "21  MEAN_ALIGNED_READ_LENGTH       220.49       \n",
       "22  READS_ALIGNED_IN_PAIRS          0           \n",
       "23  PCT_READS_ALIGNED_IN_PAIRS      0           \n",
       "24  PF_READS_IMPROPER_PAIRS         0           \n",
       "25  PCT_PF_READS_IMPROPER_PAIRS     0           \n",
       "26  BAD_CYCLES                      0           \n",
       "27  STRAND_BALANCE                 50.02%       \n",
       "28  PCT_CHIMERAS                   1.58%        \n",
       "29  PCT_ADAPTER                    0.00%        \n",
       "30  PCT_SOFTCLIP                   1.15%        \n",
       "31  PCT_HARDCLIP                    0           \n",
       "32  AVG_POS_3PRIME_SOFTCLIP_LENGTH 60.01        \n",
       "33  SAMPLE                          NaN         \n",
       "34  LIBRARY                         NaN         \n",
       "35  READ_GROUP                      NaN         "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<br><br><b>Metric type: DuplicationMetrics</b><br>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>key</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LIBRARY</td>\n",
       "      <td>L3511</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>UNPAIRED_READS_EXAMINED</td>\n",
       "      <td>607981005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>READ_PAIRS_EXAMINED</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SECONDARY_OR_SUPPLEMENTARY_RDS</td>\n",
       "      <td>10355345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UNMAPPED_READS</td>\n",
       "      <td>6350995</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>UNPAIRED_READ_DUPLICATES</td>\n",
       "      <td>39965875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>READ_PAIR_DUPLICATES</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>READ_PAIR_OPTICAL_DUPLICATES</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>PERCENT_DUPLICATION</td>\n",
       "      <td>6.57%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ESTIMATED_LIBRARY_SIZE</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              key      value\n",
       "0  LIBRARY                         L3511    \n",
       "1  UNPAIRED_READS_EXAMINED         607981005\n",
       "2  READ_PAIRS_EXAMINED             0        \n",
       "3  SECONDARY_OR_SUPPLEMENTARY_RDS  10355345 \n",
       "4  UNMAPPED_READS                  6350995  \n",
       "5  UNPAIRED_READ_DUPLICATES        39965875 \n",
       "6  READ_PAIR_DUPLICATES            0        \n",
       "7  READ_PAIR_OPTICAL_DUPLICATES    0        \n",
       "8  PERCENT_DUPLICATION            6.57%     \n",
       "9  ESTIMATED_LIBRARY_SIZE          NaN      "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<br><br><b>Metric type: GcBiasDetailMetrics</b><br>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>key</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACCUMULATION_LEVEL</td>\n",
       "      <td>All Reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>READS_USED</td>\n",
       "      <td>ALL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GC</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WINDOWS</td>\n",
       "      <td>136305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>READ_STARTS</td>\n",
       "      <td>16041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MEAN_BASE_QUALITY</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NORMALIZED_COVERAGE</td>\n",
       "      <td>58.04%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ERROR_BAR_WIDTH</td>\n",
       "      <td>0.46%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SAMPLE</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>LIBRARY</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>READ_GROUP</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    key      value\n",
       "0   ACCUMULATION_LEVEL   All Reads\n",
       "1   READS_USED           ALL      \n",
       "2   GC                   0        \n",
       "3   WINDOWS              136305   \n",
       "4   READ_STARTS          16041    \n",
       "5   MEAN_BASE_QUALITY    14       \n",
       "6   NORMALIZED_COVERAGE 58.04%    \n",
       "7   ERROR_BAR_WIDTH     0.46%     \n",
       "8   SAMPLE               NaN      \n",
       "9   LIBRARY              NaN      \n",
       "10  READ_GROUP           NaN      "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<br><br><b>Metric type: GcBiasSummaryMetrics</b><br>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>key</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACCUMULATION_LEVEL</td>\n",
       "      <td>All Reads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>READS_USED</td>\n",
       "      <td>ALL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WINDOW_SIZE</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TOTAL_CLUSTERS</td>\n",
       "      <td>624687345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ALIGNED_READS</td>\n",
       "      <td>618336350</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AT_DROPOUT</td>\n",
       "      <td>5.45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GC_DROPOUT</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GC_NC_0_19</td>\n",
       "      <td>52.85%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GC_NC_20_39</td>\n",
       "      <td>89.49%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>GC_NC_40_59</td>\n",
       "      <td>1.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>GC_NC_60_79</td>\n",
       "      <td>1.35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>GC_NC_80_100</td>\n",
       "      <td>1.82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>SAMPLE</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>LIBRARY</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>READ_GROUP</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   key      value\n",
       "0   ACCUMULATION_LEVEL  All Reads\n",
       "1   READS_USED          ALL      \n",
       "2   WINDOW_SIZE         100      \n",
       "3   TOTAL_CLUSTERS      624687345\n",
       "4   ALIGNED_READS       618336350\n",
       "5   AT_DROPOUT         5.45      \n",
       "6   GC_DROPOUT         0.00%     \n",
       "7   GC_NC_0_19         52.85%    \n",
       "8   GC_NC_20_39        89.49%    \n",
       "9   GC_NC_40_59        1.08      \n",
       "10  GC_NC_60_79        1.35      \n",
       "11  GC_NC_80_100       1.82      \n",
       "12  SAMPLE              NaN      \n",
       "13  LIBRARY             NaN      \n",
       "14  READ_GROUP          NaN      "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<br><br><b>Metric type: QualityYieldMetricsFlow</b><br>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>key</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>READ_LENGTH_AVG_Q_ABOVE_30</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>READ_LENGTH_AVG_Q_ABOVE_25</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TOTAL_READS</td>\n",
       "      <td>614332000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PF_READS</td>\n",
       "      <td>612938364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>READ_LENGTH</td>\n",
       "      <td>224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TOTAL_BASES</td>\n",
       "      <td>137807226292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PF_BASES</td>\n",
       "      <td>137717547905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Q20_BASES</td>\n",
       "      <td>119903460651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>PF_Q20_BASES</td>\n",
       "      <td>119828940593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Q30_BASES</td>\n",
       "      <td>91082044488</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>PF_Q30_BASES</td>\n",
       "      <td>91026891229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Q20_EQUIVALENT_YIELD</td>\n",
       "      <td>223909781529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>PF_Q20_EQUIVALENT_YIELD</td>\n",
       "      <td>223769400848</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           key         value\n",
       "0   READ_LENGTH_AVG_Q_ABOVE_30  0           \n",
       "1   READ_LENGTH_AVG_Q_ABOVE_25  0           \n",
       "2   TOTAL_READS                 614332000   \n",
       "3   PF_READS                    612938364   \n",
       "4   READ_LENGTH                 224         \n",
       "5   TOTAL_BASES                 137807226292\n",
       "6   PF_BASES                    137717547905\n",
       "7   Q20_BASES                   119903460651\n",
       "8   PF_Q20_BASES                119828940593\n",
       "9   Q30_BASES                   91082044488 \n",
       "10  PF_Q30_BASES                91026891229 \n",
       "11  Q20_EQUIVALENT_YIELD        223909781529\n",
       "12  PF_Q20_EQUIVALENT_YIELD     223769400848"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<br><br><b>Metric type: RawWgsMetrics</b><br>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>key</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GENOME_TERRITORY</td>\n",
       "      <td>2,937,690,299.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MEAN_COVERAGE</td>\n",
       "      <td>41.79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SD_COVERAGE</td>\n",
       "      <td>57.44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MEDIAN_COVERAGE</td>\n",
       "      <td>40.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MAD_COVERAGE</td>\n",
       "      <td>8.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PCT_EXC_ADAPTER</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PCT_EXC_MAPQ</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PCT_EXC_DUPE</td>\n",
       "      <td>5.87%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>PCT_EXC_UNPAIRED</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>PCT_EXC_BASEQ</td>\n",
       "      <td>0.14%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>PCT_EXC_OVERLAP</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>PCT_EXC_CAPPED</td>\n",
       "      <td>0.07%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>PCT_EXC_TOTAL</td>\n",
       "      <td>6.09%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>PCT_1X</td>\n",
       "      <td>99.20%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>PCT_5X</td>\n",
       "      <td>98.63%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>PCT_10X</td>\n",
       "      <td>97.92%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>PCT_15X</td>\n",
       "      <td>96.87%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>PCT_20X</td>\n",
       "      <td>94.83%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>PCT_25X</td>\n",
       "      <td>90.43%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>PCT_30X</td>\n",
       "      <td>81.83%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>PCT_40X</td>\n",
       "      <td>52.42%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>PCT_50X</td>\n",
       "      <td>23.84%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>PCT_60X</td>\n",
       "      <td>8.50%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>PCT_70X</td>\n",
       "      <td>2.75%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>PCT_80X</td>\n",
       "      <td>1.06%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>PCT_90X</td>\n",
       "      <td>0.63%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>PCT_100X</td>\n",
       "      <td>0.48%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>FOLD_80_BASE_PENALTY</td>\n",
       "      <td>1.39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>FOLD_90_BASE_PENALTY</td>\n",
       "      <td>1.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>FOLD_95_BASE_PENALTY</td>\n",
       "      <td>2.20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>HET_SNP_SENSITIVITY</td>\n",
       "      <td>98.81%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>HET_SNP_Q</td>\n",
       "      <td>19.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     key            value\n",
       "0   GENOME_TERRITORY     2,937,690,299.00\n",
       "1   MEAN_COVERAGE        41.79           \n",
       "2   SD_COVERAGE          57.44           \n",
       "3   MEDIAN_COVERAGE      40.00           \n",
       "4   MAD_COVERAGE         8.00            \n",
       "5   PCT_EXC_ADAPTER      0.00%           \n",
       "6   PCT_EXC_MAPQ         0.00%           \n",
       "7   PCT_EXC_DUPE         5.87%           \n",
       "8   PCT_EXC_UNPAIRED     0.00%           \n",
       "9   PCT_EXC_BASEQ        0.14%           \n",
       "10  PCT_EXC_OVERLAP      0.00%           \n",
       "11  PCT_EXC_CAPPED       0.07%           \n",
       "12  PCT_EXC_TOTAL        6.09%           \n",
       "13  PCT_1X               99.20%          \n",
       "14  PCT_5X               98.63%          \n",
       "15  PCT_10X              97.92%          \n",
       "16  PCT_15X              96.87%          \n",
       "17  PCT_20X              94.83%          \n",
       "18  PCT_25X              90.43%          \n",
       "19  PCT_30X              81.83%          \n",
       "20  PCT_40X              52.42%          \n",
       "21  PCT_50X              23.84%          \n",
       "22  PCT_60X              8.50%           \n",
       "23  PCT_70X              2.75%           \n",
       "24  PCT_80X              1.06%           \n",
       "25  PCT_90X              0.63%           \n",
       "26  PCT_100X             0.48%           \n",
       "27  FOLD_80_BASE_PENALTY 1.39            \n",
       "28  FOLD_90_BASE_PENALTY 1.67            \n",
       "29  FOLD_95_BASE_PENALTY 2.20            \n",
       "30  HET_SNP_SENSITIVITY  98.81%          \n",
       "31  HET_SNP_Q            19.00           "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<br><br><b>Metric type: WgsMetrics</b><br>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>key</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GENOME_TERRITORY</td>\n",
       "      <td>2937690299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MEAN_COVERAGE</td>\n",
       "      <td>33.91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SD_COVERAGE</td>\n",
       "      <td>26.09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MEDIAN_COVERAGE</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MAD_COVERAGE</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PCT_EXC_ADAPTER</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PCT_EXC_MAPQ</td>\n",
       "      <td>8.63%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PCT_EXC_DUPE</td>\n",
       "      <td>4.81%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>PCT_EXC_UNPAIRED</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>PCT_EXC_BASEQ</td>\n",
       "      <td>10.36%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>PCT_EXC_OVERLAP</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>PCT_EXC_CAPPED</td>\n",
       "      <td>0.01%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>PCT_EXC_TOTAL</td>\n",
       "      <td>23.82%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>PCT_1X</td>\n",
       "      <td>94.43%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>PCT_5X</td>\n",
       "      <td>93.21%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>PCT_10X</td>\n",
       "      <td>92.04%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>PCT_15X</td>\n",
       "      <td>90.08%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>PCT_20X</td>\n",
       "      <td>86.09%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>PCT_25X</td>\n",
       "      <td>78.34%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>PCT_30X</td>\n",
       "      <td>65.86%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>PCT_40X</td>\n",
       "      <td>34.47%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>PCT_50X</td>\n",
       "      <td>12.40%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>PCT_60X</td>\n",
       "      <td>3.37%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>PCT_70X</td>\n",
       "      <td>0.76%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>PCT_80X</td>\n",
       "      <td>0.19%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>PCT_90X</td>\n",
       "      <td>0.09%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>PCT_100X</td>\n",
       "      <td>0.07%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>FOLD_80_BASE_PENALTY</td>\n",
       "      <td>1.41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>FOLD_90_BASE_PENALTY</td>\n",
       "      <td>2.26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>FOLD_95_BASE_PENALTY</td>\n",
       "      <td>?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>HET_SNP_SENSITIVITY</td>\n",
       "      <td>93.92%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>HET_SNP_Q</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     key       value\n",
       "0   GENOME_TERRITORY      2937690299\n",
       "1   MEAN_COVERAGE        33.91      \n",
       "2   SD_COVERAGE          26.09      \n",
       "3   MEDIAN_COVERAGE       35        \n",
       "4   MAD_COVERAGE          8         \n",
       "5   PCT_EXC_ADAPTER      0.00%      \n",
       "6   PCT_EXC_MAPQ         8.63%      \n",
       "7   PCT_EXC_DUPE         4.81%      \n",
       "8   PCT_EXC_UNPAIRED      0         \n",
       "9   PCT_EXC_BASEQ        10.36%     \n",
       "10  PCT_EXC_OVERLAP       0         \n",
       "11  PCT_EXC_CAPPED       0.01%      \n",
       "12  PCT_EXC_TOTAL        23.82%     \n",
       "13  PCT_1X               94.43%     \n",
       "14  PCT_5X               93.21%     \n",
       "15  PCT_10X              92.04%     \n",
       "16  PCT_15X              90.08%     \n",
       "17  PCT_20X              86.09%     \n",
       "18  PCT_25X              78.34%     \n",
       "19  PCT_30X              65.86%     \n",
       "20  PCT_40X              34.47%     \n",
       "21  PCT_50X              12.40%     \n",
       "22  PCT_60X              3.37%      \n",
       "23  PCT_70X              0.76%      \n",
       "24  PCT_80X              0.19%      \n",
       "25  PCT_90X              0.09%      \n",
       "26  PCT_100X             0.07%      \n",
       "27  FOLD_80_BASE_PENALTY 1.41       \n",
       "28  FOLD_90_BASE_PENALTY 2.26       \n",
       "29  FOLD_95_BASE_PENALTY  ?         \n",
       "30  HET_SNP_SENSITIVITY  93.92%     \n",
       "31  HET_SNP_Q             12        "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<br><br><b>Metric type: stats_coverage</b><br>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>key</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Genome___percentile_5</td>\n",
       "      <td>21.51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Genome___percentile_10</td>\n",
       "      <td>26.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Genome___median_coverage</td>\n",
       "      <td>41.29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Genome___median_coverage_normalized</td>\n",
       "      <td>100.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Genome___percent_larger_than_05_of_genome_median</td>\n",
       "      <td>96.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Genome___percent_larger_than_025_of_genome_median</td>\n",
       "      <td>98.99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Genome___percent_over_or_equal_to_10x</td>\n",
       "      <td>99.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Genome___percent_over_or_equal_to_20x</td>\n",
       "      <td>96.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Exome (WG)___percentile_5</td>\n",
       "      <td>25.92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Exome (WG)___percentile_10</td>\n",
       "      <td>31.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Exome (WG)___median_coverage</td>\n",
       "      <td>50.06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Exome (WG)___median_coverage_normalized</td>\n",
       "      <td>1.21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Exome (WG)___percent_larger_than_05_of_genome_median</td>\n",
       "      <td>97.57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Exome (WG)___percent_larger_than_025_of_genome_median</td>\n",
       "      <td>98.98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Exome (WG)___percent_over_or_equal_to_10x</td>\n",
       "      <td>99.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Exome (WG)___percent_over_or_equal_to_20x</td>\n",
       "      <td>97.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ACMG59 (WG)___percentile_5</td>\n",
       "      <td>28.38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ACMG59 (WG)___percentile_10</td>\n",
       "      <td>32.65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ACMG59 (WG)___median_coverage</td>\n",
       "      <td>49.84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ACMG59 (WG)___median_coverage_normalized</td>\n",
       "      <td>1.21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ACMG59 (WG)___percent_larger_than_05_of_genome_median</td>\n",
       "      <td>98.99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ACMG59 (WG)___percent_larger_than_025_of_genome_median</td>\n",
       "      <td>99.97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ACMG59 (WG)___percent_over_or_equal_to_10x</td>\n",
       "      <td>99.98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ACMG59 (WG)___percent_over_or_equal_to_20x</td>\n",
       "      <td>99.27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Exon 1___percentile_5</td>\n",
       "      <td>31.66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Exon 1___percentile_10</td>\n",
       "      <td>37.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Exon 1___median_coverage</td>\n",
       "      <td>58.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Exon 1___median_coverage_normalized</td>\n",
       "      <td>1.41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Exon 1___percent_larger_than_05_of_genome_median</td>\n",
       "      <td>99.31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Exon 1___percent_larger_than_025_of_genome_median</td>\n",
       "      <td>99.89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Exon 1___percent_over_or_equal_to_10x</td>\n",
       "      <td>99.91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Exon 1___percent_over_or_equal_to_20x</td>\n",
       "      <td>99.43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Exon &gt;1___percentile_5</td>\n",
       "      <td>26.38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Exon &gt;1___percentile_10</td>\n",
       "      <td>30.52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Exon &gt;1___median_coverage</td>\n",
       "      <td>47.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Exon &gt;1___median_coverage_normalized</td>\n",
       "      <td>1.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Exon &gt;1___percent_larger_than_05_of_genome_median</td>\n",
       "      <td>98.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Exon &gt;1___percent_larger_than_025_of_genome_median</td>\n",
       "      <td>99.78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Exon &gt;1___percent_over_or_equal_to_10x</td>\n",
       "      <td>99.82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Exon &gt;1___percent_over_or_equal_to_20x</td>\n",
       "      <td>98.75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>Regulatory (WG)___percentile_5</td>\n",
       "      <td>25.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>Regulatory (WG)___percentile_10</td>\n",
       "      <td>30.36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>Regulatory (WG)___median_coverage</td>\n",
       "      <td>45.87</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>Regulatory (WG)___median_coverage_normalized</td>\n",
       "      <td>1.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>Regulatory (WG)___percent_larger_than_05_of_genome_median</td>\n",
       "      <td>98.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>Regulatory (WG)___percent_larger_than_025_of_genome_median</td>\n",
       "      <td>99.46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>Regulatory (WG)___percent_over_or_equal_to_10x</td>\n",
       "      <td>99.51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>Regulatory (WG)___percent_over_or_equal_to_20x</td>\n",
       "      <td>98.31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>OMIM___percentile_5</td>\n",
       "      <td>21.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>OMIM___percentile_10</td>\n",
       "      <td>26.27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>OMIM___median_coverage</td>\n",
       "      <td>40.63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>OMIM___median_coverage_normalized</td>\n",
       "      <td>98.41%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>OMIM___percent_larger_than_05_of_genome_median</td>\n",
       "      <td>96.34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>OMIM___percent_larger_than_025_of_genome_median</td>\n",
       "      <td>98.98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>OMIM___percent_over_or_equal_to_10x</td>\n",
       "      <td>99.07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>OMIM___percent_over_or_equal_to_20x</td>\n",
       "      <td>96.86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>GC 10-20___percentile_5</td>\n",
       "      <td>65.22%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>GC 10-20___percentile_10</td>\n",
       "      <td>4.54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>GC 10-20___median_coverage</td>\n",
       "      <td>26.39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>GC 10-20___median_coverage_normalized</td>\n",
       "      <td>63.93%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>GC 10-20___percent_larger_than_05_of_genome_median</td>\n",
       "      <td>70.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>GC 10-20___percent_larger_than_025_of_genome_median</td>\n",
       "      <td>84.53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>GC 10-20___percent_over_or_equal_to_10x</td>\n",
       "      <td>85.51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>GC 10-20___percent_over_or_equal_to_20x</td>\n",
       "      <td>72.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>GC 20-30___percentile_5</td>\n",
       "      <td>17.44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>GC 20-30___percentile_10</td>\n",
       "      <td>21.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>GC 20-30___median_coverage</td>\n",
       "      <td>33.59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>GC 20-30___median_coverage_normalized</td>\n",
       "      <td>81.35%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>GC 20-30___percent_larger_than_05_of_genome_median</td>\n",
       "      <td>92.39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>GC 20-30___percent_larger_than_025_of_genome_median</td>\n",
       "      <td>98.38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>GC 20-30___percent_over_or_equal_to_10x</td>\n",
       "      <td>98.59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>GC 20-30___percent_over_or_equal_to_20x</td>\n",
       "      <td>93.57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>GC 30-40___percentile_5</td>\n",
       "      <td>19.99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>GC 30-40___percentile_10</td>\n",
       "      <td>24.52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>GC 30-40___median_coverage</td>\n",
       "      <td>37.99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>GC 30-40___median_coverage_normalized</td>\n",
       "      <td>92.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>GC 30-40___percent_larger_than_05_of_genome_median</td>\n",
       "      <td>94.99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>GC 30-40___percent_larger_than_025_of_genome_median</td>\n",
       "      <td>98.75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>GC 30-40___percent_over_or_equal_to_10x</td>\n",
       "      <td>98.92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>GC 30-40___percent_over_or_equal_to_20x</td>\n",
       "      <td>95.68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>GC 40-50___percentile_5</td>\n",
       "      <td>24.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>GC 40-50___percentile_10</td>\n",
       "      <td>28.77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>GC 40-50___median_coverage</td>\n",
       "      <td>43.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>GC 40-50___median_coverage_normalized</td>\n",
       "      <td>1.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>GC 40-50___percent_larger_than_05_of_genome_median</td>\n",
       "      <td>97.38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>GC 40-50___percent_larger_than_025_of_genome_median</td>\n",
       "      <td>99.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>GC 40-50___percent_over_or_equal_to_10x</td>\n",
       "      <td>99.49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>GC 40-50___percent_over_or_equal_to_20x</td>\n",
       "      <td>97.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>GC 50-60___percentile_5</td>\n",
       "      <td>28.55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>GC 50-60___percentile_10</td>\n",
       "      <td>33.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>GC 50-60___median_coverage</td>\n",
       "      <td>49.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>GC 50-60___median_coverage_normalized</td>\n",
       "      <td>1.19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>GC 50-60___percent_larger_than_05_of_genome_median</td>\n",
       "      <td>98.65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>GC 50-60___percent_larger_than_025_of_genome_median</td>\n",
       "      <td>99.71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>GC 50-60___percent_over_or_equal_to_10x</td>\n",
       "      <td>99.76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>GC 50-60___percent_over_or_equal_to_20x</td>\n",
       "      <td>98.85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>GC 60-70___percentile_5</td>\n",
       "      <td>32.87</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>GC 60-70___percentile_10</td>\n",
       "      <td>38.66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>GC 60-70___median_coverage</td>\n",
       "      <td>57.52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>GC 60-70___median_coverage_normalized</td>\n",
       "      <td>1.39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>GC 60-70___percent_larger_than_05_of_genome_median</td>\n",
       "      <td>98.92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>GC 60-70___percent_larger_than_025_of_genome_median</td>\n",
       "      <td>99.70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>GC 60-70___percent_over_or_equal_to_10x</td>\n",
       "      <td>99.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>GC 60-70___percent_over_or_equal_to_20x</td>\n",
       "      <td>99.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>GC 70-80___percentile_5</td>\n",
       "      <td>29.58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>GC 70-80___percentile_10</td>\n",
       "      <td>39.93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>GC 70-80___median_coverage</td>\n",
       "      <td>62.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>GC 70-80___median_coverage_normalized</td>\n",
       "      <td>1.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>GC 70-80___percent_larger_than_05_of_genome_median</td>\n",
       "      <td>97.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>GC 70-80___percent_larger_than_025_of_genome_median</td>\n",
       "      <td>99.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>GC 70-80___percent_over_or_equal_to_10x</td>\n",
       "      <td>99.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>GC 70-80___percent_over_or_equal_to_20x</td>\n",
       "      <td>97.60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>GC 80-90___percentile_5</td>\n",
       "      <td>16.21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>GC 80-90___percentile_10</td>\n",
       "      <td>24.32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>GC 80-90___median_coverage</td>\n",
       "      <td>59.62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>GC 80-90___median_coverage_normalized</td>\n",
       "      <td>1.44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>GC 80-90___percent_larger_than_05_of_genome_median</td>\n",
       "      <td>92.58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>GC 80-90___percent_larger_than_025_of_genome_median</td>\n",
       "      <td>98.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>GC 80-90___percent_over_or_equal_to_10x</td>\n",
       "      <td>98.51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>GC 80-90___percent_over_or_equal_to_20x</td>\n",
       "      <td>93.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>GC 90-100___percentile_5</td>\n",
       "      <td>12.24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>GC 90-100___percentile_10</td>\n",
       "      <td>22.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>GC 90-100___median_coverage</td>\n",
       "      <td>51.49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>GC 90-100___median_coverage_normalized</td>\n",
       "      <td>1.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>GC 90-100___percent_larger_than_05_of_genome_median</td>\n",
       "      <td>90.92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>GC 90-100___percent_larger_than_025_of_genome_median</td>\n",
       "      <td>95.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>GC 90-100___percent_over_or_equal_to_10x</td>\n",
       "      <td>96.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>GC 90-100___percent_over_or_equal_to_20x</td>\n",
       "      <td>91.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>Unique___percentile_5</td>\n",
       "      <td>22.26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>Unique___percentile_10</td>\n",
       "      <td>26.48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>Unique___median_coverage</td>\n",
       "      <td>41.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>Unique___median_coverage_normalized</td>\n",
       "      <td>99.85%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>Unique___percent_larger_than_05_of_genome_median</td>\n",
       "      <td>96.59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>Unique___percent_larger_than_025_of_genome_median</td>\n",
       "      <td>99.28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>Unique___percent_over_or_equal_to_10x</td>\n",
       "      <td>99.38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>Unique___percent_over_or_equal_to_20x</td>\n",
       "      <td>97.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>Non-unique___percentile_5</td>\n",
       "      <td>9.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>Non-unique___percentile_10</td>\n",
       "      <td>16.49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>Non-unique___median_coverage</td>\n",
       "      <td>42.75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>Non-unique___median_coverage_normalized</td>\n",
       "      <td>1.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>Non-unique___percent_larger_than_05_of_genome_median</td>\n",
       "      <td>87.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>Non-unique___percent_larger_than_025_of_genome_median</td>\n",
       "      <td>94.41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>Non-unique___percent_over_or_equal_to_10x</td>\n",
       "      <td>95.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>Non-unique___percent_over_or_equal_to_20x</td>\n",
       "      <td>87.98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>DNA repeats___percentile_5</td>\n",
       "      <td>22.43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>DNA repeats___percentile_10</td>\n",
       "      <td>26.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>DNA repeats___median_coverage</td>\n",
       "      <td>39.57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>DNA repeats___median_coverage_normalized</td>\n",
       "      <td>95.84%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>DNA repeats___percent_larger_than_05_of_genome_median</td>\n",
       "      <td>96.94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>DNA repeats___percent_larger_than_025_of_genome_median</td>\n",
       "      <td>99.60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>DNA repeats___percent_over_or_equal_to_10x</td>\n",
       "      <td>99.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>DNA repeats___percent_over_or_equal_to_20x</td>\n",
       "      <td>97.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>LINE___percentile_5</td>\n",
       "      <td>21.69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>LINE___percentile_10</td>\n",
       "      <td>25.57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>LINE___median_coverage</td>\n",
       "      <td>39.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>LINE___median_coverage_normalized</td>\n",
       "      <td>94.74%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>LINE___percent_larger_than_05_of_genome_median</td>\n",
       "      <td>96.38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>LINE___percent_larger_than_025_of_genome_median</td>\n",
       "      <td>99.44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>LINE___percent_over_or_equal_to_10x</td>\n",
       "      <td>99.54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>LINE___percent_over_or_equal_to_20x</td>\n",
       "      <td>97.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>LTR___percentile_5</td>\n",
       "      <td>25.72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>LTR___percentile_10</td>\n",
       "      <td>29.92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>LTR___median_coverage</td>\n",
       "      <td>44.79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163</th>\n",
       "      <td>LTR___median_coverage_normalized</td>\n",
       "      <td>1.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>LTR___percent_larger_than_05_of_genome_median</td>\n",
       "      <td>98.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>LTR___percent_larger_than_025_of_genome_median</td>\n",
       "      <td>99.63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>LTR___percent_over_or_equal_to_10x</td>\n",
       "      <td>99.69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>LTR___percent_over_or_equal_to_20x</td>\n",
       "      <td>98.45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>Low complexity___percentile_5</td>\n",
       "      <td>16.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>Low complexity___percentile_10</td>\n",
       "      <td>20.89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>Low complexity___median_coverage</td>\n",
       "      <td>37.79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>Low complexity___median_coverage_normalized</td>\n",
       "      <td>91.53%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>Low complexity___percent_larger_than_05_of_genome_median</td>\n",
       "      <td>91.21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>Low complexity___percent_larger_than_025_of_genome_median</td>\n",
       "      <td>97.97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>Low complexity___percent_over_or_equal_to_10x</td>\n",
       "      <td>98.27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>Low complexity___percent_over_or_equal_to_20x</td>\n",
       "      <td>92.35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>SINE___percentile_5</td>\n",
       "      <td>24.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td>SINE___percentile_10</td>\n",
       "      <td>28.49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>SINE___median_coverage</td>\n",
       "      <td>42.51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179</th>\n",
       "      <td>SINE___median_coverage_normalized</td>\n",
       "      <td>1.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>SINE___percent_larger_than_05_of_genome_median</td>\n",
       "      <td>97.42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>SINE___percent_larger_than_025_of_genome_median</td>\n",
       "      <td>99.47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>SINE___percent_over_or_equal_to_10x</td>\n",
       "      <td>99.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>SINE___percent_over_or_equal_to_20x</td>\n",
       "      <td>97.80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184</th>\n",
       "      <td>Alu repeats___percentile_5</td>\n",
       "      <td>24.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>Alu repeats___percentile_10</td>\n",
       "      <td>28.61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>Alu repeats___median_coverage</td>\n",
       "      <td>42.62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>Alu repeats___median_coverage_normalized</td>\n",
       "      <td>1.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>Alu repeats___percent_larger_than_05_of_genome_median</td>\n",
       "      <td>97.27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>Alu repeats___percent_larger_than_025_of_genome_median</td>\n",
       "      <td>99.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>Alu repeats___percent_over_or_equal_to_10x</td>\n",
       "      <td>99.49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>Alu repeats___percent_over_or_equal_to_20x</td>\n",
       "      <td>97.65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192</th>\n",
       "      <td>Satellite___percentile_5</td>\n",
       "      <td>4.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>Satellite___percentile_10</td>\n",
       "      <td>7.88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>Satellite___median_coverage</td>\n",
       "      <td>39.22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>Satellite___median_coverage_normalized</td>\n",
       "      <td>94.99%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>Satellite___percent_larger_than_05_of_genome_median</td>\n",
       "      <td>74.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>Satellite___percent_larger_than_025_of_genome_median</td>\n",
       "      <td>87.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>Satellite___percent_over_or_equal_to_10x</td>\n",
       "      <td>88.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>Satellite___percent_over_or_equal_to_20x</td>\n",
       "      <td>75.44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>Simple repeat___percentile_5</td>\n",
       "      <td>45.17%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201</th>\n",
       "      <td>Simple repeat___percentile_10</td>\n",
       "      <td>4.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202</th>\n",
       "      <td>Simple repeat___median_coverage</td>\n",
       "      <td>30.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>Simple repeat___median_coverage_normalized</td>\n",
       "      <td>73.62%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204</th>\n",
       "      <td>Simple repeat___percent_larger_than_05_of_genome_median</td>\n",
       "      <td>69.35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>Simple repeat___percent_larger_than_025_of_genome_median</td>\n",
       "      <td>83.28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>206</th>\n",
       "      <td>Simple repeat___percent_over_or_equal_to_10x</td>\n",
       "      <td>84.47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207</th>\n",
       "      <td>Simple repeat___percent_over_or_equal_to_20x</td>\n",
       "      <td>70.99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>LCR___percentile_5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209</th>\n",
       "      <td>LCR___percentile_10</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210</th>\n",
       "      <td>LCR___median_coverage</td>\n",
       "      <td>24.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211</th>\n",
       "      <td>LCR___median_coverage_normalized</td>\n",
       "      <td>58.50%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>LCR___percent_larger_than_05_of_genome_median</td>\n",
       "      <td>55.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>LCR___percent_larger_than_025_of_genome_median</td>\n",
       "      <td>62.69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214</th>\n",
       "      <td>LCR___percent_over_or_equal_to_10x</td>\n",
       "      <td>63.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215</th>\n",
       "      <td>LCR___percent_over_or_equal_to_20x</td>\n",
       "      <td>56.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216</th>\n",
       "      <td>Non-LCR___percentile_5</td>\n",
       "      <td>22.06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>Non-LCR___percentile_10</td>\n",
       "      <td>26.43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>Non-LCR___median_coverage</td>\n",
       "      <td>41.44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>Non-LCR___median_coverage_normalized</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>Non-LCR___percent_larger_than_05_of_genome_median</td>\n",
       "      <td>96.41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221</th>\n",
       "      <td>Non-LCR___percent_larger_than_025_of_genome_median</td>\n",
       "      <td>99.21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>222</th>\n",
       "      <td>Non-LCR___percent_over_or_equal_to_10x</td>\n",
       "      <td>99.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>223</th>\n",
       "      <td>Non-LCR___percent_over_or_equal_to_20x</td>\n",
       "      <td>96.93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>224</th>\n",
       "      <td>Strong_sec_struct___percentile_5</td>\n",
       "      <td>1.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>Strong_sec_struct___percentile_10</td>\n",
       "      <td>3.72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>226</th>\n",
       "      <td>Strong_sec_struct___median_coverage</td>\n",
       "      <td>19.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227</th>\n",
       "      <td>Strong_sec_struct___median_coverage_normalized</td>\n",
       "      <td>46.37%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>228</th>\n",
       "      <td>Strong_sec_struct___percent_larger_than_05_of_genome_median</td>\n",
       "      <td>47.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>229</th>\n",
       "      <td>Strong_sec_struct___percent_larger_than_025_of_genome_median</td>\n",
       "      <td>76.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>Strong_sec_struct___percent_over_or_equal_to_10x</td>\n",
       "      <td>78.58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231</th>\n",
       "      <td>Strong_sec_struct___percent_over_or_equal_to_20x</td>\n",
       "      <td>50.45</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                              key   value\n",
       "0    Genome___percentile_5                                        21.51  \n",
       "1    Genome___percentile_10                                       26.10  \n",
       "2    Genome___median_coverage                                     41.29  \n",
       "3    Genome___median_coverage_normalized                          100.00%\n",
       "4    Genome___percent_larger_than_05_of_genome_median             96.02  \n",
       "5    Genome___percent_larger_than_025_of_genome_median            98.99  \n",
       "6    Genome___percent_over_or_equal_to_10x                        99.12  \n",
       "7    Genome___percent_over_or_equal_to_20x                        96.56  \n",
       "8    Exome (WG)___percentile_5                                    25.92  \n",
       "9    Exome (WG)___percentile_10                                   31.14  \n",
       "10   Exome (WG)___median_coverage                                 50.06  \n",
       "11   Exome (WG)___median_coverage_normalized                      1.21   \n",
       "12   Exome (WG)___percent_larger_than_05_of_genome_median         97.57  \n",
       "13   Exome (WG)___percent_larger_than_025_of_genome_median        98.98  \n",
       "14   Exome (WG)___percent_over_or_equal_to_10x                    99.08  \n",
       "15   Exome (WG)___percent_over_or_equal_to_20x                    97.81  \n",
       "16   ACMG59 (WG)___percentile_5                                   28.38  \n",
       "17   ACMG59 (WG)___percentile_10                                  32.65  \n",
       "18   ACMG59 (WG)___median_coverage                                49.84  \n",
       "19   ACMG59 (WG)___median_coverage_normalized                     1.21   \n",
       "20   ACMG59 (WG)___percent_larger_than_05_of_genome_median        98.99  \n",
       "21   ACMG59 (WG)___percent_larger_than_025_of_genome_median       99.97  \n",
       "22   ACMG59 (WG)___percent_over_or_equal_to_10x                   99.98  \n",
       "23   ACMG59 (WG)___percent_over_or_equal_to_20x                   99.27  \n",
       "24   Exon 1___percentile_5                                        31.66  \n",
       "25   Exon 1___percentile_10                                       37.10  \n",
       "26   Exon 1___median_coverage                                     58.04  \n",
       "27   Exon 1___median_coverage_normalized                          1.41   \n",
       "28   Exon 1___percent_larger_than_05_of_genome_median             99.31  \n",
       "29   Exon 1___percent_larger_than_025_of_genome_median            99.89  \n",
       "30   Exon 1___percent_over_or_equal_to_10x                        99.91  \n",
       "31   Exon 1___percent_over_or_equal_to_20x                        99.43  \n",
       "32   Exon >1___percentile_5                                       26.38  \n",
       "33   Exon >1___percentile_10                                      30.52  \n",
       "34   Exon >1___median_coverage                                    47.33  \n",
       "35   Exon >1___median_coverage_normalized                         1.15   \n",
       "36   Exon >1___percent_larger_than_05_of_genome_median            98.50  \n",
       "37   Exon >1___percent_larger_than_025_of_genome_median           99.78  \n",
       "38   Exon >1___percent_over_or_equal_to_10x                       99.82  \n",
       "39   Exon >1___percent_over_or_equal_to_20x                       98.75  \n",
       "40   Regulatory (WG)___percentile_5                               25.90  \n",
       "41   Regulatory (WG)___percentile_10                              30.36  \n",
       "42   Regulatory (WG)___median_coverage                            45.87  \n",
       "43   Regulatory (WG)___median_coverage_normalized                 1.11   \n",
       "44   Regulatory (WG)___percent_larger_than_05_of_genome_median    98.04  \n",
       "45   Regulatory (WG)___percent_larger_than_025_of_genome_median   99.46  \n",
       "46   Regulatory (WG)___percent_over_or_equal_to_10x               99.51  \n",
       "47   Regulatory (WG)___percent_over_or_equal_to_20x               98.31  \n",
       "48   OMIM___percentile_5                                          21.96  \n",
       "49   OMIM___percentile_10                                         26.27  \n",
       "50   OMIM___median_coverage                                       40.63  \n",
       "51   OMIM___median_coverage_normalized                            98.41% \n",
       "52   OMIM___percent_larger_than_05_of_genome_median               96.34  \n",
       "53   OMIM___percent_larger_than_025_of_genome_median              98.98  \n",
       "54   OMIM___percent_over_or_equal_to_10x                          99.07  \n",
       "55   OMIM___percent_over_or_equal_to_20x                          96.86  \n",
       "56   GC 10-20___percentile_5                                      65.22% \n",
       "57   GC 10-20___percentile_10                                     4.54   \n",
       "58   GC 10-20___median_coverage                                   26.39  \n",
       "59   GC 10-20___median_coverage_normalized                        63.93% \n",
       "60   GC 10-20___percent_larger_than_05_of_genome_median           70.15  \n",
       "61   GC 10-20___percent_larger_than_025_of_genome_median          84.53  \n",
       "62   GC 10-20___percent_over_or_equal_to_10x                      85.51  \n",
       "63   GC 10-20___percent_over_or_equal_to_20x                      72.33  \n",
       "64   GC 20-30___percentile_5                                      17.44  \n",
       "65   GC 20-30___percentile_10                                     21.56  \n",
       "66   GC 20-30___median_coverage                                   33.59  \n",
       "67   GC 20-30___median_coverage_normalized                        81.35% \n",
       "68   GC 20-30___percent_larger_than_05_of_genome_median           92.39  \n",
       "69   GC 20-30___percent_larger_than_025_of_genome_median          98.38  \n",
       "70   GC 20-30___percent_over_or_equal_to_10x                      98.59  \n",
       "71   GC 20-30___percent_over_or_equal_to_20x                      93.57  \n",
       "72   GC 30-40___percentile_5                                      19.99  \n",
       "73   GC 30-40___percentile_10                                     24.52  \n",
       "74   GC 30-40___median_coverage                                   37.99  \n",
       "75   GC 30-40___median_coverage_normalized                        92.00% \n",
       "76   GC 30-40___percent_larger_than_05_of_genome_median           94.99  \n",
       "77   GC 30-40___percent_larger_than_025_of_genome_median          98.75  \n",
       "78   GC 30-40___percent_over_or_equal_to_10x                      98.92  \n",
       "79   GC 30-40___percent_over_or_equal_to_20x                      95.68  \n",
       "80   GC 40-50___percentile_5                                      24.25  \n",
       "81   GC 40-50___percentile_10                                     28.77  \n",
       "82   GC 40-50___median_coverage                                   43.14  \n",
       "83   GC 40-50___median_coverage_normalized                        1.04   \n",
       "84   GC 40-50___percent_larger_than_05_of_genome_median           97.38  \n",
       "85   GC 40-50___percent_larger_than_025_of_genome_median          99.40  \n",
       "86   GC 40-50___percent_over_or_equal_to_10x                      99.49  \n",
       "87   GC 40-50___percent_over_or_equal_to_20x                      97.74  \n",
       "88   GC 50-60___percentile_5                                      28.55  \n",
       "89   GC 50-60___percentile_10                                     33.23  \n",
       "90   GC 50-60___median_coverage                                   49.23  \n",
       "91   GC 50-60___median_coverage_normalized                        1.19   \n",
       "92   GC 50-60___percent_larger_than_05_of_genome_median           98.65  \n",
       "93   GC 50-60___percent_larger_than_025_of_genome_median          99.71  \n",
       "94   GC 50-60___percent_over_or_equal_to_10x                      99.76  \n",
       "95   GC 50-60___percent_over_or_equal_to_20x                      98.85  \n",
       "96   GC 60-70___percentile_5                                      32.87  \n",
       "97   GC 60-70___percentile_10                                     38.66  \n",
       "98   GC 60-70___median_coverage                                   57.52  \n",
       "99   GC 60-70___median_coverage_normalized                        1.39   \n",
       "100  GC 60-70___percent_larger_than_05_of_genome_median           98.92  \n",
       "101  GC 60-70___percent_larger_than_025_of_genome_median          99.70  \n",
       "102  GC 60-70___percent_over_or_equal_to_10x                      99.74  \n",
       "103  GC 60-70___percent_over_or_equal_to_20x                      99.04  \n",
       "104  GC 70-80___percentile_5                                      29.58  \n",
       "105  GC 70-80___percentile_10                                     39.93  \n",
       "106  GC 70-80___median_coverage                                   62.00  \n",
       "107  GC 70-80___median_coverage_normalized                        1.50   \n",
       "108  GC 70-80___percent_larger_than_05_of_genome_median           97.40  \n",
       "109  GC 70-80___percent_larger_than_025_of_genome_median          99.04  \n",
       "110  GC 70-80___percent_over_or_equal_to_10x                      99.16  \n",
       "111  GC 70-80___percent_over_or_equal_to_20x                      97.60  \n",
       "112  GC 80-90___percentile_5                                      16.21  \n",
       "113  GC 80-90___percentile_10                                     24.32  \n",
       "114  GC 80-90___median_coverage                                   59.62  \n",
       "115  GC 80-90___median_coverage_normalized                        1.44   \n",
       "116  GC 80-90___percent_larger_than_05_of_genome_median           92.58  \n",
       "117  GC 80-90___percent_larger_than_025_of_genome_median          98.13  \n",
       "118  GC 80-90___percent_over_or_equal_to_10x                      98.51  \n",
       "119  GC 80-90___percent_over_or_equal_to_20x                      93.23  \n",
       "120  GC 90-100___percentile_5                                     12.24  \n",
       "121  GC 90-100___percentile_10                                    22.40  \n",
       "122  GC 90-100___median_coverage                                  51.49  \n",
       "123  GC 90-100___median_coverage_normalized                       1.25   \n",
       "124  GC 90-100___percent_larger_than_05_of_genome_median          90.92  \n",
       "125  GC 90-100___percent_larger_than_025_of_genome_median         95.81  \n",
       "126  GC 90-100___percent_over_or_equal_to_10x                     96.00  \n",
       "127  GC 90-100___percent_over_or_equal_to_20x                     91.12  \n",
       "128  Unique___percentile_5                                        22.26  \n",
       "129  Unique___percentile_10                                       26.48  \n",
       "130  Unique___median_coverage                                     41.23  \n",
       "131  Unique___median_coverage_normalized                          99.85% \n",
       "132  Unique___percent_larger_than_05_of_genome_median             96.59  \n",
       "133  Unique___percent_larger_than_025_of_genome_median            99.28  \n",
       "134  Unique___percent_over_or_equal_to_10x                        99.38  \n",
       "135  Unique___percent_over_or_equal_to_20x                        97.11  \n",
       "136  Non-unique___percentile_5                                    9.05   \n",
       "137  Non-unique___percentile_10                                   16.49  \n",
       "138  Non-unique___median_coverage                                 42.75  \n",
       "139  Non-unique___median_coverage_normalized                      1.04   \n",
       "140  Non-unique___percent_larger_than_05_of_genome_median         87.11  \n",
       "141  Non-unique___percent_larger_than_025_of_genome_median        94.41  \n",
       "142  Non-unique___percent_over_or_equal_to_10x                    95.03  \n",
       "143  Non-unique___percent_over_or_equal_to_20x                    87.98  \n",
       "144  DNA repeats___percentile_5                                   22.43  \n",
       "145  DNA repeats___percentile_10                                  26.18  \n",
       "146  DNA repeats___median_coverage                                39.57  \n",
       "147  DNA repeats___median_coverage_normalized                     95.84% \n",
       "148  DNA repeats___percent_larger_than_05_of_genome_median        96.94  \n",
       "149  DNA repeats___percent_larger_than_025_of_genome_median       99.60  \n",
       "150  DNA repeats___percent_over_or_equal_to_10x                   99.67  \n",
       "151  DNA repeats___percent_over_or_equal_to_20x                   97.50  \n",
       "152  LINE___percentile_5                                          21.69  \n",
       "153  LINE___percentile_10                                         25.57  \n",
       "154  LINE___median_coverage                                       39.12  \n",
       "155  LINE___median_coverage_normalized                            94.74% \n",
       "156  LINE___percent_larger_than_05_of_genome_median               96.38  \n",
       "157  LINE___percent_larger_than_025_of_genome_median              99.44  \n",
       "158  LINE___percent_over_or_equal_to_10x                          99.54  \n",
       "159  LINE___percent_over_or_equal_to_20x                          97.01  \n",
       "160  LTR___percentile_5                                           25.72  \n",
       "161  LTR___percentile_10                                          29.92  \n",
       "162  LTR___median_coverage                                        44.79  \n",
       "163  LTR___median_coverage_normalized                             1.08   \n",
       "164  LTR___percent_larger_than_05_of_genome_median                98.16  \n",
       "165  LTR___percent_larger_than_025_of_genome_median               99.63  \n",
       "166  LTR___percent_over_or_equal_to_10x                           99.69  \n",
       "167  LTR___percent_over_or_equal_to_20x                           98.45  \n",
       "168  Low complexity___percentile_5                                16.10  \n",
       "169  Low complexity___percentile_10                               20.89  \n",
       "170  Low complexity___median_coverage                             37.79  \n",
       "171  Low complexity___median_coverage_normalized                  91.53% \n",
       "172  Low complexity___percent_larger_than_05_of_genome_median     91.21  \n",
       "173  Low complexity___percent_larger_than_025_of_genome_median    97.97  \n",
       "174  Low complexity___percent_over_or_equal_to_10x                98.27  \n",
       "175  Low complexity___percent_over_or_equal_to_20x                92.35  \n",
       "176  SINE___percentile_5                                          24.11  \n",
       "177  SINE___percentile_10                                         28.49  \n",
       "178  SINE___median_coverage                                       42.51  \n",
       "179  SINE___median_coverage_normalized                            1.03   \n",
       "180  SINE___percent_larger_than_05_of_genome_median               97.42  \n",
       "181  SINE___percent_larger_than_025_of_genome_median              99.47  \n",
       "182  SINE___percent_over_or_equal_to_10x                          99.56  \n",
       "183  SINE___percent_over_or_equal_to_20x                          97.80  \n",
       "184  Alu repeats___percentile_5                                   24.04  \n",
       "185  Alu repeats___percentile_10                                  28.61  \n",
       "186  Alu repeats___median_coverage                                42.62  \n",
       "187  Alu repeats___median_coverage_normalized                     1.03   \n",
       "188  Alu repeats___percent_larger_than_05_of_genome_median        97.27  \n",
       "189  Alu repeats___percent_larger_than_025_of_genome_median       99.40  \n",
       "190  Alu repeats___percent_over_or_equal_to_10x                   99.49  \n",
       "191  Alu repeats___percent_over_or_equal_to_20x                   97.65  \n",
       "192  Satellite___percentile_5                                     4.14   \n",
       "193  Satellite___percentile_10                                    7.88   \n",
       "194  Satellite___median_coverage                                  39.22  \n",
       "195  Satellite___median_coverage_normalized                       94.99% \n",
       "196  Satellite___percent_larger_than_05_of_genome_median          74.15  \n",
       "197  Satellite___percent_larger_than_025_of_genome_median         87.16  \n",
       "198  Satellite___percent_over_or_equal_to_10x                     88.50  \n",
       "199  Satellite___percent_over_or_equal_to_20x                     75.44  \n",
       "200  Simple repeat___percentile_5                                 45.17% \n",
       "201  Simple repeat___percentile_10                                4.12   \n",
       "202  Simple repeat___median_coverage                              30.40  \n",
       "203  Simple repeat___median_coverage_normalized                   73.62% \n",
       "204  Simple repeat___percent_larger_than_05_of_genome_median      69.35  \n",
       "205  Simple repeat___percent_larger_than_025_of_genome_median     83.28  \n",
       "206  Simple repeat___percent_over_or_equal_to_10x                 84.47  \n",
       "207  Simple repeat___percent_over_or_equal_to_20x                 70.99  \n",
       "208  LCR___percentile_5                                           0.00%  \n",
       "209  LCR___percentile_10                                          0.00%  \n",
       "210  LCR___median_coverage                                        24.16  \n",
       "211  LCR___median_coverage_normalized                             58.50% \n",
       "212  LCR___percent_larger_than_05_of_genome_median                55.16  \n",
       "213  LCR___percent_larger_than_025_of_genome_median               62.69  \n",
       "214  LCR___percent_over_or_equal_to_10x                           63.23  \n",
       "215  LCR___percent_over_or_equal_to_20x                           56.17  \n",
       "216  Non-LCR___percentile_5                                       22.06  \n",
       "217  Non-LCR___percentile_10                                      26.43  \n",
       "218  Non-LCR___median_coverage                                    41.44  \n",
       "219  Non-LCR___median_coverage_normalized                         1.00   \n",
       "220  Non-LCR___percent_larger_than_05_of_genome_median            96.41  \n",
       "221  Non-LCR___percent_larger_than_025_of_genome_median           99.21  \n",
       "222  Non-LCR___percent_over_or_equal_to_10x                       99.33  \n",
       "223  Non-LCR___percent_over_or_equal_to_20x                       96.93  \n",
       "224  Strong_sec_struct___percentile_5                             1.12   \n",
       "225  Strong_sec_struct___percentile_10                            3.72   \n",
       "226  Strong_sec_struct___median_coverage                          19.14  \n",
       "227  Strong_sec_struct___median_coverage_normalized               46.37% \n",
       "228  Strong_sec_struct___percent_larger_than_05_of_genome_median  47.33  \n",
       "229  Strong_sec_struct___percent_larger_than_025_of_genome_median 76.15  \n",
       "230  Strong_sec_struct___percent_over_or_equal_to_10x             78.58  \n",
       "231  Strong_sec_struct___percent_over_or_equal_to_20x             50.45  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "to_print = pd.DataFrame()\n",
    "sorted_keys  = tbl_list_keys['metric'].sort_values()\n",
    "\n",
    "for tbl_key in sorted_keys:\n",
    "    \n",
    "    HTML(\"<br>\"+\"<br>\"+\"<b>\"+\"Metric type: \"+tbl_key+\"</b>\"+\"<br>\")\n",
    "    df_h5_tbl = pd.read_hdf(input_h5_file, tbl_key).T\n",
    "    df_h5_tbl = df_h5_tbl.reset_index()\n",
    "    \n",
    "\n",
    "    if (tbl_key.startswith('stats') ):\n",
    "        df_h5_tbl['metric'] = df_h5_tbl['level_0'] + separator + df_h5_tbl['level_1']\n",
    "        df_h5_tbl = df_h5_tbl.drop(columns=['level_0','level_1']).copy()\n",
    "        df_h5_tbl.columns = ['value','key']\n",
    "        to_print = df_h5_tbl[['key', 'value']]        \n",
    "        to_print\n",
    "\n",
    "    else:\n",
    "        to_print = df_h5_tbl\n",
    "        to_print.columns = ['key','value'] \n",
    "        to_print\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "# FIN #"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Genomics",
   "language": "python",
   "name": "genomics"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
