{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Detailed Variant Report v0.3\n",
    "\n",
    "## Data Input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import h5py\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from configparser import ConfigParser\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "pd.options.display.float_format = '{:,.2%}'.format\n",
    "\n",
    "configFile='detailed_var_report.config'\n",
    "parser = ConfigParser()\n",
    "parser.read(configFile)\n",
    "\n",
    "prmNames=['run_id','pipeline_version',\n",
    "          'h5_concordance_file']\n",
    "prm={}\n",
    "for name in prmNames:\n",
    "    prm[name]=parser.get('VarReport', name, fallback='NA')\n",
    "\n",
    "h5outfile = parser.get('VarReport', 'h5_output', fallback='detailed_var_report.h5')\n",
    "    \n",
    "source = prm['h5_concordance_file']\n",
    "\n",
    "chromData=[]\n",
    "\n",
    "chrKeys=['/chr{:d}'.format(i+1) for i in range(22)]\n",
    "\n",
    "for k in chrKeys:\n",
    "    d=pd.read_hdf(source, key=k, mode='r')\n",
    "    chromData.append(d)\n",
    "\n",
    "#print([len(chromData[i]) for i in range(len(chromData))])\n",
    "\n",
    "data=pd.concat(chromData)\n",
    "data['tree_score'] = np.where(data['classify']=='fn',-1,data['tree_score'])\n",
    "#print(len(data))\n",
    "chromData=[]\n",
    "\n",
    "prm['mean_var_depth']='{:.2f}'.format(data['well_mapped_coverage'].mean())\n",
    "prmNames.append('mean_var_depth')\n",
    "   \n",
    "try:\n",
    "    args=pd.read_hdf(sources[0], 'input_args', mode='r')\n",
    "    prm['truth_sample_name']=args['truth_sample_name'][0]\n",
    "except:\n",
    "    prm['truth_sample_name']=parser.get('VarReport', 'truth_sample_name', fallback='NA')\n",
    "prmNames.append('truth_sample_name')\n",
    "\n",
    "prm['regions']=parser.get('DetailedReport', 'regions').split(',')\n",
    "prmNames.append('regions')\n",
    "\n",
    "\n",
    "prmdf = pd.DataFrame.from_dict(prm, orient='index',columns=['value']).reindex(prmNames)\n",
    "prmdf.to_hdf(h5outfile, key=\"det_parameters\")\n",
    "prmdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data['label'] = np.where(data['classify']=='fp',0,1)\n",
    "data['filter'][data['filter']=='HPOL_RUN']='PASS'\n",
    "data['filter'][data['classify']=='fn']='MISS'\n",
    "\n",
    "data['class'] = np.where(data['label']==0,'FP','FN')\n",
    "data['class'][(data['label']==1) & (data['filter']=='PASS')]='TP'\n",
    "data['class'][(data['label']==0) & (data['filter']!='PASS')]='TN'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filterByRegion(data,region):\n",
    "    if region=='All':\n",
    "        return data\n",
    "    elif region.startswith('Non-'):\n",
    "        return data[data[region[4:]]==False]\n",
    "    else:\n",
    "        return data[data[region]==True]\n",
    "\n",
    "def filterByRange(data,prm,mn,mx):\n",
    "    return data[(data[prm]>=mn) & (data[prm]<mx)]\n",
    "\n",
    "def filterByVal(data,prm,val):\n",
    "    return data[(data[prm]==val)]\n",
    "\n",
    "def filterByVarCategory(data,cat):\n",
    "    if cat=='All':\n",
    "        return data\n",
    "    if cat=='SNP':\n",
    "        return data[data['indel']==False]\n",
    "    elif cat=='cycleskip SNP':\n",
    "        return data[(data['indel']==False) & (data['cycleskip_status'])==True]\n",
    "    elif cat=='Indel':\n",
    "        return data[data['indel']==True]\n",
    "    elif cat=='non-hmer':\n",
    "            return data[(data['indel']==True) & (data['hmer_indel_length']==0) & (data['indel_length']>1)]\n",
    "    elif cat=='hmer 0-1':\n",
    "        return data[(data['indel']==True) & (data['hmer_indel_length']==0) & (data['indel_length']<=1)]\n",
    "    elif cat=='hmer 2-4':\n",
    "        return data[(data['indel']==True) & (data['hmer_indel_length']>0) & (data['hmer_indel_length']<=4)]\n",
    "    elif cat=='hmer 5-8':\n",
    "        return data[(data['indel']==True) & (data['hmer_indel_length']>4) & (data['hmer_indel_length']<=8)]\n",
    "    elif cat=='hmer 9-10':\n",
    "        return data[(data['indel']==True) & (data['hmer_indel_length']>8) & (data['hmer_indel_length']<=10)]\n",
    "    elif cat=='hmer 11+':\n",
    "        return data[(data['indel']==True) & (data['hmer_indel_length']>10)]\n",
    "    return None\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calcPerformanceStatic(d):\n",
    "    \n",
    "    num=len(d)\n",
    "    numPos=sum(d['label'])\n",
    "    numNeg=num-numPos\n",
    "    if num==0 or numPos==0:\n",
    "        return (None,numPos,numNeg)\n",
    "    \n",
    "    fn=len(d[d['class']=='FN'])\n",
    "    tp=len(d[d['class']=='TP'])\n",
    "    fp=len(d[d['class']=='FP'])\n",
    "    \n",
    "    recall=tp/(tp+fn) if (tp+fn>0) else np.nan\n",
    "    precision=tp/(tp+fp) if (tp+fp>0) else np.nan\n",
    "    max_recall=1-len(d[d['filter']=='MISS'])/numPos\n",
    "    \n",
    "    f1=tp/(tp+0.5*fn+0.5*fp)\n",
    "\n",
    "    return ({'max_recall':max_recall,'recall':recall,'precision':precision,'f1':f1,'tp':tp,'fp':fp,'fn':fn},numPos,numNeg)\n",
    "\n",
    "\n",
    "def calcPerformanceOptimized(d1):\n",
    "    \n",
    "    d=d1[['classify','tree_score','label']].sort_values(by=['tree_score']).copy()\n",
    "    \n",
    "    num=len(d)\n",
    "    numPos=sum(d['label'])\n",
    "    numNeg=num-numPos\n",
    "    if num==0:\n",
    "        return (None,numPos,numNeg)\n",
    "    \n",
    "    d['fn']=np.cumsum(d['label'])\n",
    "    d['tp']=numPos-(d['fn'])\n",
    "    d['fp']=numNeg-np.cumsum(1-d['label'])\n",
    "\n",
    "    d['recall']=d['tp']/(d['tp']+d['fn'])\n",
    "    d['precision']=d['tp']/(d['tp']+d['fp'])\n",
    "\n",
    "    d['f1']=d['tp']/(d['tp']+0.5*d['fn']+0.5*d['fp'])\n",
    "\n",
    "    d['mask']=((d['tp']+d['fn'])>=1) & ((d['tp']+d['fp'])>=1) & (d['tree_score']>=0)\n",
    "    if len(d[d['mask']])==0:\n",
    "        return (None,numPos,numNeg)\n",
    "    \n",
    "    maxF1=max(d[d['mask']]['f1'])\n",
    "    d['opt']=(d['f1']==maxF1)\n",
    "\n",
    "    r=(d[d['opt']])\n",
    "    #print(r.iloc[0,:].to_dict())\n",
    "    return (r.iloc[0,:].to_dict(),numPos,numNeg)\n",
    "    #return ({'recall':r['recall'],'precision':r['precision'],'f1':r['f1'],'tp':r['tp'],'fp':r['fp'],'fn':r['fn']},numPos,numNeg)\n",
    "\n",
    "\n",
    "def renameReg(reg):\n",
    "    regRename={\"LCR-hs38\": \"LCR\", \"Non-LCR-hs38\": \"Non-LCR\", \"exome.twist\": \"Exome\"}\n",
    "    return regRename[reg] if reg in regRename else reg\n",
    "    \n",
    "\n",
    "def getStats(d,reg,cat,var):\n",
    "    resStat,pos,neg=calcPerformanceStatic(d)\n",
    "    resOpt,pos1,neg1=calcPerformanceOptimized(d)\n",
    "    mi=pd.MultiIndex.from_tuples([(renameReg(reg),var,cat)],names=['Region','Variant','Category'])\n",
    "\n",
    "    row=pd.DataFrame({'# pos':pos,\n",
    "                      '# neg':neg,\n",
    "                      'TP':np.nan if not resStat else resStat['tp'],\n",
    "                      'FP':np.nan if not resStat else resStat['fp'],\n",
    "                      'FN':np.nan if not resStat else resStat['fn'],\n",
    "                      'max recall':np.nan if not resStat else resStat['max_recall'],\n",
    "                      'recall-stat':np.nan if not resStat else resStat['recall'],\n",
    "                      'precision-stat':np.nan if not resStat else resStat['precision'],\n",
    "                      'F1-stat':np.nan if not resStat else resStat['f1'],\n",
    "                      'recall-opt':np.nan if not resOpt else resOpt['recall'],\n",
    "                      'precision-opt':np.nan if not resOpt else resOpt['precision'],\n",
    "                      'F1-opt':np.nan if not resOpt else resOpt['f1'],\n",
    "                      'avg cvg':d['coverage'].mean(),\n",
    "                      'avg gc':d['gc_content'].mean(),\n",
    "                     },index=mi)\n",
    "    \n",
    "    #print(resOpt)\n",
    "    \n",
    "    return row\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Summary Performance - Genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.options.display.float_format = lambda x: '{:,.2%}'.format(x) if x<=1 and x>0 else \\\n",
    "                                           ('{:,.0f}'.format(x) if x.is_integer() else \\\n",
    "                                            '{:,.1f}'.format(x))\n",
    "\n",
    "varCats=['All','SNP','Indel','non-hmer','hmer 0-1','hmer 2-4','hmer 5-8','hmer 9-10','hmer 11+']\n",
    "#varCats=['SNP','Indel']\n",
    "#varCats=['hmer Indel >8,<=10','hmer Indel >10']\n",
    "\n",
    "\n",
    "res=pd.DataFrame()\n",
    "for region in ['All']+prm['regions']:\n",
    "#for region in ['exome.twist']:\n",
    "    d1=filterByRegion(data,region)\n",
    "    \n",
    "    for var in varCats:\n",
    "        #print(region,var)\n",
    "        d2=filterByVarCategory(d1,var)\n",
    "        res=pd.concat([res,getStats(d2,region, 'All', var)])\n",
    "        \n",
    "        for gc in [(0,0.2),(0.2,0.8),(0.8,1)]:\n",
    "            d3=filterByRange(d2,'gc_content',gc[0],gc[1])\n",
    "            res=pd.concat([res,getStats(d3,region, f'GC {gc[0]*100:.0f}-{gc[1]*100:.0f}', var)])\n",
    "        for cvg in [(0,20),(20,40),(40,100)]:\n",
    "            d3=filterByRange(d2,'coverage',cvg[0],cvg[1])\n",
    "            res=pd.concat([res,getStats(d3,region, 'CVG {0}-{1}'.format(cvg[0],cvg[1]), var)])\n",
    "        for lcr in [True,False]:\n",
    "            d3=filterByVal(d2,'LCR-hs38',lcr)\n",
    "            res=pd.concat([res,getStats(d3,region, 'LCR' if lcr else 'Non-LCR', var)])\n",
    "        for mappab in [True,False]:\n",
    "            d3=filterByVal(d2,'mappability.0',mappab)\n",
    "            res=pd.concat([res,getStats(d3,region, 'Mappable' if mappab else 'Non-Mappable'.format(mappab), var)])\n",
    "       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "res.to_hdf(h5outfile, key=\"detailed_vars\")\n",
    "\n",
    "parser.read(configFile)\n",
    "csvFile=parser.get('DetailedReport', 'csv')\n",
    "\n",
    "out=res.reset_index()\n",
    "out.to_csv(csvFile,index=False)\n",
    "#out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "columns = varCats #[:3]\n",
    "rows = ['All', 'GC 0-20', 'GC 20-80', 'GC 80-100', 'LCR', 'Non-LCR', 'CVG 0-20', 'CVG 20-40', 'CVG 40-100', \n",
    "        'Mappable','Non-Mappable','Exome'] #[:3]\n",
    "\n",
    "def getRow(cat,var,exome=False):\n",
    "    reg='Exome' if exome else 'All'\n",
    "    x=out[((((out['Category']==reg) & (out['Region']==cat)) | \n",
    "            ((out['Category']==cat) & (out['Region']==reg))) & \n",
    "           (out['Variant']==var))]\n",
    "    return x\n",
    "\n",
    "def getTabTxt(cat,var,prm,exome=False):\n",
    "    x=getRow(cat,var,exome)\n",
    "    n=x['# pos'].iloc[0] \n",
    "    num='{:d}k'.format(int(n/1000)) if n>1000 else '{:d}'.format(n)\n",
    "    return '{:.1%}\\n({:s},{:.1f})'.format(x[prm].iloc[0],num,x['avg cvg'].iloc[0])\n",
    "\n",
    "def getTabCol(cat,var,prm,exome=False):\n",
    "    x=getRow(cat,var,exome)\n",
    "    if x['# pos'].iloc[0]<30:\n",
    "        return 'white'\n",
    "    return plt.cm.RdYlGn(x[prm].iloc[0])\n",
    "\n",
    "\n",
    "tabl = [[getTabTxt(r,c,'max recall') for c in columns] for r in rows] \n",
    "tabcol = [[getTabCol(r,c,'max recall') for c in columns] for r in rows] \n",
    "\n",
    "fig, ax = plt.subplots(figsize=(20,10)) \n",
    "ax.set_axis_off() \n",
    "table = ax.table( \n",
    "    cellText = tabl,  \n",
    "    rowLabels = rows,  \n",
    "    colLabels = columns,\n",
    "    cellColours = tabcol,\n",
    "    cellLoc ='center',  \n",
    "    loc ='upper left')     \n",
    "table.set_fontsize(30)\n",
    "table.scale(1, 4)\n",
    "x=plt.title('Genome - VC performance per category: max recall (n,cvg)', fontsize=30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "tabl = [[getTabTxt(r,c,'F1-stat') for c in columns] for r in rows] \n",
    "tabcol = [[getTabCol(r,c,'F1-stat') for c in columns] for r in rows] \n",
    "\n",
    "fig, ax = plt.subplots(figsize=(20,10)) \n",
    "ax.set_axis_off() \n",
    "table = ax.table( \n",
    "    cellText = tabl,  \n",
    "    rowLabels = rows,  \n",
    "    colLabels = columns,\n",
    "    cellColours = tabcol,\n",
    "    cellLoc ='center',  \n",
    "    loc ='upper left')     \n",
    "table.set_fontsize(30)\n",
    "table.scale(1, 4)\n",
    "x=plt.title('Genome - VC performance per category: F1 (n,cvg)', fontsize=30)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "tabl = [[getTabTxt(r,c,'F1-opt') for c in columns] for r in rows] \n",
    "tabcol = [[getTabCol(r,c,'F1-opt') for c in columns] for r in rows] \n",
    "\n",
    "fig, ax = plt.subplots(figsize=(20,10)) \n",
    "ax.set_axis_off() \n",
    "table = ax.table( \n",
    "    cellText = tabl,  \n",
    "    rowLabels = rows,  \n",
    "    colLabels = columns,\n",
    "    cellColours = tabcol,\n",
    "    cellLoc ='center',  \n",
    "    loc ='upper left')     \n",
    "table.set_fontsize(30)\n",
    "table.scale(1, 4)\n",
    "x=plt.title('Genome - VC performance per category: Reoptimized F1 (n,cvg)', fontsize=30)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Summary Performance - Exome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "columns = varCats #[:3]\n",
    "rows = ['All', 'GC 0-20', 'GC 20-80', 'GC 80-100', 'LCR', 'Non-LCR', 'CVG 0-20', 'CVG 20-40', 'CVG 40-100', \n",
    "        'Mappable','Non-Mappable'] #[:3]\n",
    "\n",
    "tabl = [[getTabTxt(r,c,'max recall',True) for c in columns] for r in rows] \n",
    "tabcol = [[getTabCol(r,c,'max recall',True) for c in columns] for r in rows] \n",
    "\n",
    "fig, ax = plt.subplots(figsize=(20,10)) \n",
    "ax.set_axis_off() \n",
    "table = ax.table( \n",
    "    cellText = tabl,  \n",
    "    rowLabels = rows,  \n",
    "    colLabels = columns,\n",
    "    cellColours = tabcol,\n",
    "    cellLoc ='center',  \n",
    "    loc ='upper left')     \n",
    "table.set_fontsize(30)\n",
    "table.scale(1, 4)\n",
    "x=plt.title('Exome - VC performance per category: max recall (n,cvg)', fontsize=30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tabl = [[getTabTxt(r,c,'F1-stat',True) for c in columns] for r in rows] \n",
    "tabcol = [[getTabCol(r,c,'F1-stat',True) for c in columns] for r in rows] \n",
    "\n",
    "fig, ax = plt.subplots(figsize=(20,10)) \n",
    "ax.set_axis_off() \n",
    "table = ax.table( \n",
    "    cellText = tabl,  \n",
    "    rowLabels = rows,  \n",
    "    colLabels = columns,\n",
    "    cellColours = tabcol,\n",
    "    cellLoc ='center',  \n",
    "    loc ='upper left')     \n",
    "table.set_fontsize(30)\n",
    "table.scale(1, 4)\n",
    "x=plt.title('Exome - VC performance per category: F1 (n,cvg)', fontsize=30)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tabl = [[getTabTxt(r,c,'F1-opt',True) for c in columns] for r in rows] \n",
    "tabcol = [[getTabCol(r,c,'F1-opt',True) for c in columns] for r in rows] \n",
    "\n",
    "fig, ax = plt.subplots(figsize=(20,10)) \n",
    "ax.set_axis_off() \n",
    "table = ax.table( \n",
    "    cellText = tabl,  \n",
    "    rowLabels = rows,  \n",
    "    colLabels = columns,\n",
    "    cellColours = tabcol,\n",
    "    cellLoc ='center',  \n",
    "    loc ='upper left')     \n",
    "table.set_fontsize(30)\n",
    "table.scale(1, 4)\n",
    "x=plt.title('Exome - VC performance per category: Optimized F1 (n,cvg)', fontsize=30)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Detailed Performance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_rows', 1000)\n",
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.8"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "384px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
